File: solutions/c/rna-transcription/1/rna_transcription.c
#include <stdlib.h>
#include <string.h>

#include "rna_transcription.h"

char *to_rna(const char *dna)
{
    char *rna = (char*)malloc(strlen(dna) + 1);
    if (rna == NULL)
        return NULL;
    
    char *rna_start = rna;
    char n;
    while ((n = *dna++)) {
        switch (n) {
            case    'G' : *rna++ = 'C';
            break;
            case    'C' : *rna++ = 'G';
            break;
            case    'T' : *rna++ = 'A';
            break;
            case    'A' : *rna++ = 'U';
            break;
            default :
                *rna = '\0';
                return rna_start;
        }
    }
    *rna = n;
    return rna_start;
}

File: solutions/go/rna-transcription/1/rna_transcription.go
package strand

var transcription = map[rune]rune{
	'G': 'C',
	'C': 'G',
	'T': 'A',
	'A': 'U',
}

func ToRNA(dna string) string {
	var rna string
	for _, n := range dna {
		if t, ok := transcription[n]; ok {
			rna += string(t)
		} else {
			break
		} // Stop processing on invalid nucleotide
	}
	return rna
}
